src/ngraph/pass/concat_fusion.hpp

```diff
@@ -0,0 +1,60 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include "ngraph/pass/graph_rewrite.hpp"
+#include "ngraph/pass/pass.hpp"
+#include "ngraph/pattern/matcher.hpp"
+#include "ngraph/pattern/op/label.hpp"
+
+namespace ngraph
+{
+    namespace pass
+    {
+        class ConcatElimination;
+        class SelfConcatFusion;
+    }
+}
+
+class ngraph::pass::ConcatElimination : public ngraph::pass::GraphRewrite
+{
+public:
+    ConcatElimination()
+        : GraphRewrite()
+    {
+        construct_concat_elimination();
+    }
+
+private:
+    void construct_concat_elimination();
+};
+
+class ngraph::pass::SelfConcatFusion : public ngraph::pass::FunctionPass
+{
+public:
+    virtual bool run_on_function(std::shared_ptr<ngraph::Function> function) override;
+
+private:
+    void update_concat_pattern_vectors(const std::shared_ptr<Node>&, size_t);
+    void remove_single_concat_op_pattern();
+    void construct_concat_patterns(const std::shared_ptr<pattern::Matcher>&,
+                                   const std::shared_ptr<pattern::op::Label>&,
+                                   const std::shared_ptr<Node>&);
+    bool replace_patterns(const std::pair<NodeVector, std::vector<size_t>>&);
+
+    std::vector<std::pair<NodeVector, std::vector<size_t>>> m_concat_pattern_vectors;
```

Comment:
Instead of keep all the self concat in a NodeVector, can we just keep the last `Concat` in the vector, or unordered_set, which can simplify the code a lot. The parent concat nodes can be easily deduced from the last one.

Reply:
I do need the first and the last concat of the pattern plus all their concat axes. I can create a something that will just store the first and last at each iteration where the match happens, but I thought it might be easier to just have the nodevector. Based on this design, I definitely need the argument to the first concat and the last concat to replace the pattern. I thought it is simpler to keep track of the concat nodes as a set. Please let me know what you think.

Reply:
Just feel this data structure is too deeply nested, and therefore less straightforward to understand what you like to do.
Since it is necessary to keep the whole concat node vector, do we still need the `second` in the pair? Cannot it be derived from the node by `get_concatenation_axis()`?

Reply:
Yes..I can be derived from the `NodeVector` using another loop. This is what I was going to do. ButI figured why not do it in one go and bind it to the corresponding `NodeVector`. Do you want me change the implementation or is ok to keep it this way?

Reply:
Made the changes...Please take a look